### ISSUE TITLE
Fix several corner cases (in branch exp-space)

### ISSUE THREAD

**Issue by GiggleLiu** [under-Peter/OMEinsum.jl#4]
Branch: exp-space

## When output is empty
```julia
julia> einsum(((1, 2), (2, 1)), (a, a), ())
ERROR: MethodError: Base.reduce_empty(::typeof(Base.add_sum), ::Type{Union{}}) is ambiguous. Candidates:
  reduce_empty(::typeof(Base.add_sum), ::Type{T}) where T<:Union{UInt16, UInt32, UInt8} in Base at reduce.jl:236
  reduce_empty(::typeof(Base.add_sum), ::Type{T}) where T<:Union{Int16, Int32, Int8} in Base at reduce.jl:235
Possible fix, define
  reduce_empty(::typeof(Base.add_sum), ::Type{Union{}})
Stacktrace:
 [1] mapreduce_empty(::typeof(identity), ::Function, ::Type) at ./reduce.jl:251
 [2] mapreduce_empty_iter(::Function, ::Function, ::Tuple{}, ::Base.HasEltype) at ./reduce.jl:258
 [3] mapfoldl_impl at ./reduce.jl:57 [inlined]
 [4] #mapfoldl#187 at ./reduce.jl:72 [inlined]
 [5] mapfoldl at ./reduce.jl:72 [inlined]
 [6] #mapreduce#191 at ./reduce.jl:205 [inlined]
 [7] mapreduce(::Function, ::Function, ::Tuple{}) at ./reduce.jl:205
 [8] sum(::Function, ::Tuple{}) at ./reduce.jl:399
 [9] sum(::Tuple{}) at ./reduce.jl:416
 [10] macro expansion at ./multidimensional.jl:1337 [inlined]
 [11] permutedims!(::Array{Float64,0}, ::Array{Float64,0}, ::Tuple{}) at ./multidimensional.jl:1328
 [12] permutedims(::Array{Float64,0}, ::Tuple{}) at ./multidimensional.jl:1312
 [13] einsum(::Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64}}, ::Tuple{Array{Float64,2},Array{Float64,2}}, ::Tuple{}) at /home/leo/jcode/OMEinsum.jl/src/einsum.jl:49
 [14] top-level scope at none:0
```

## When summing over one of the indices
```julia
julia> einsum(((1, 2), (2, 3)), (a, a), (1,))
ERROR: ArgumentError: no valid permutation of dimensions
Stacktrace:
 [1] permutedims(::Array{Float64,2}, ::Tuple{Int64}) at ./multidimensional.jl:1309
 [2] einsum(::Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64}}, ::Tuple{Array{Float64,2},Array{Float64,2}}, ::Tuple{Int64}) at /home/leo/jcode/OMEinsum.jl/src/einsum.jl:49
 [3] top-level scope at none:0
```

and this one
```julia
julia> einsum(((1,2), (1,3), (2,3,4)), (a, a, b), (1,4))
ERROR: ArgumentError: no valid permutation of dimensions
Stacktrace:
 [1] permutedims(::Array{Float64,1}, ::Tuple{Int64,Int64}) at ./multidimensional.jl:1309
 [2] einsum(::Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64},Tuple{Int64,Int64,Int64}}, ::Tuple{Array{Float64,2},Array{Float64,2},Array{Float64,3}}, ::Tuple{Int64,Int64}) at /home/leo/jcode/OMEinsum.jl/src/einsum.jl:49
 [3] top-level scope at none:0
```

**Comment by GiggleLiu** [comment]
Confirmed pytorch is able to handle the above einsums, its implementation could be found here

https://github.com/pytorch/pytorch/blob/4404762d7dd955383acee92e6f06b48144a0742e/aten/src/ATen/native/Linear.cpp#L125

**Comment by under-Peter** [comment]
Corner cases are fixed by 461dda100d582dbe5ac6d61cacdd7780ffa8e097 and 29db3e96c1671b155f5da1e5aa1a92101f396ab1 and tests for those cases have been added. (Including test_broken for features that don't work yet, i.e. trace/partial trace.
